src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/ComInterop/ComRuntimeHelpers.cs

```diff
@@ -0,0 +1,578 @@
+                {
+                    return null;
+                }
+
+                // This assert is potentially over-restrictive since COM components can behave in quite unexpected ways.
+                // However, asserting the common expected cases ensures that we find out about the unexpected scenarios, and
+                // can investigate the scenarios to ensure that there is no bug in our own code.
+                Debug.Assert(hresult == ComHresults.TYPE_E_LIBNOTREGISTERED);
+
+                Marshal.ThrowExceptionForHR(hresult);
+            }
+
+            if (typeInfoPtr == IntPtr.Zero)
+            {
+                Marshal.ThrowExceptionForHR(ComHresults.E_FAIL);
+            }
+
+            ComTypes.ITypeInfo typeInfo = null;
+            try
+            {
+                typeInfo = Marshal.GetObjectForIUnknown(typeInfoPtr) as ComTypes.ITypeInfo;
+            }
+            finally
+            {
+                Marshal.Release(typeInfoPtr);
+            }
+
+            return typeInfo;
+        }
+
+        internal static ComTypes.TYPEATTR GetTypeAttrForTypeInfo(ComTypes.ITypeInfo typeInfo)
+        {
+            IntPtr pAttrs;
+            typeInfo.GetTypeAttr(out pAttrs);
+
+            // GetTypeAttr should never return null, this is just to be safe
+            if (pAttrs == IntPtr.Zero)
+            {
+                throw Error.CannotRetrieveTypeInformation();
+            }
+
+            try
+            {
+                return (ComTypes.TYPEATTR)Marshal.PtrToStructure(pAttrs, typeof(ComTypes.TYPEATTR));
+            }
+            finally
+            {
+                typeInfo.ReleaseTypeAttr(pAttrs);
+            }
+        }
+
+        internal static ComTypes.TYPELIBATTR GetTypeAttrForTypeLib(ComTypes.ITypeLib typeLib)
+        {
+            IntPtr pAttrs;
+            typeLib.GetLibAttr(out pAttrs);
+
+            // GetTypeAttr should never return null, this is just to be safe
+            if (pAttrs == IntPtr.Zero)
+            {
+                throw Error.CannotRetrieveTypeInformation();
+            }
+
+            try
+            {
+                return (ComTypes.TYPELIBATTR)Marshal.PtrToStructure(pAttrs, typeof(ComTypes.TYPELIBATTR));
+            }
+            finally
+            {
+                typeLib.ReleaseTLibAttr(pAttrs);
+            }
+        }
+
+        public static BoundDispEvent CreateComEvent(object rcw, Guid sourceIid, int dispid)
+        {
+            return new BoundDispEvent(rcw, sourceIid, dispid);
+        }
+
+        public static DispCallable CreateDispCallable(IDispatchComObject dispatch, ComMethodDesc method)
+        {
+            return new DispCallable(dispatch, method.Name, method.DispId);
+        }
+
+        internal static MethodInfo GetGetIDispatchForObjectMethod()
+        {
+            // GetIDispatchForObject always throws a PNSE in .NET Core, so we work around it by using GetComInterfaceForObject with our IDispatch type.
```

Comment:
Would it make sense to change `GetIDispatchForObject` to forward to `GetComInterfaceForObject` in .NET Core?

Reply:
I think that makes sense. It looks like we also have `MarshalNative::GetIDispatchForObjectNative` (currently unused, I think).

I filed #33217 for it.